### ISSUE TITLE
Can't move (rename) file to /tmp/ in Linux (Mac unknown), both Neko & cpp

### ISSUE THREAD

**Issue by jcward** [HaxeFoundation/haxe#4624]
Alright, another crazy issue that's probably my own fault somehow. Using cpp or neko targets, `sys.FileSystem.rename` can't move a file to `/tmp/` on my linux machine (can't test Mac just now...) `sys.io.File.saveContents` can write a file in `/tmp/` and rename can move a file to other locations (e.g. home folder.)

Here's a test:

```
class Main {

  static function main() {

    var env = Sys.environment();
    var home = env.get("HOME");
    var fn = home+"/test.bin";

    // Clean files
    function clean() {
      if (sys.FileSystem.exists(fn)) sys.FileSystem.deleteFile(fn);
      if (sys.FileSystem.exists(fn+".2")) sys.FileSystem.deleteFile(fn+".2");
      if (sys.FileSystem.exists("/tmp/move.bin")) sys.FileSystem.deleteFile("/tmp/move.bin");
      if (sys.FileSystem.exists("/tmp/write.bin")) sys.FileSystem.deleteFile("/tmp/write.bin");
    }

    clean();

    trace("Create file: "+fn);
    sys.io.File.saveContent(fn, "A test file\n");
    if (sys.FileSystem.exists(fn)) trace("Success!"); else trace("FAIL!");

    trace("Move to home folder, works?");
    sys.FileSystem.rename(fn, fn+".2");
    if (sys.FileSystem.exists(fn+".2")) trace("Success!"); else trace("FAIL!");

    trace("Move to /tmp folder, fails...");
    try { sys.FileSystem.rename(fn+".2", "/tmp/move.bin"); } catch (e:Dynamic) { }
    if (sys.FileSystem.exists("/tmp/move.bin")) trace("Success!"); else trace("FAIL!");

    trace("Write file directly to /tmp/write.bin...");
    sys.io.File.saveContent("/tmp/write.bin", "A test file\n");
    if (sys.FileSystem.exists("/tmp/write.bin")) trace("Success!"); else trace("FAIL!");

  }
}
```

And the output on my machine:

```
Main.hx:21: Create file: /home/jward/test.bin
Main.hx:23: Success!
Main.hx:25: Move to home folder, works?
Main.hx:27: Success!
Main.hx:29: Move to /tmp folder, fails...
Main.hx:31: FAIL!
Main.hx:33: Write file directly to /tmp/write.bin...
Main.hx:35: Success!
```

FYI: this doesn't apply to Windows, there is no `/tmp/`

1) Can someone test this on Mac?
2) Can someone verify this on Linux?
3) Can someone spot my mistake? =D

**Comment by jcward** [comment]
BTW, I'm on Ubuntu 14.04, Haxe 3.2.0, neko 2.0.0, etc.

**Comment by jonasmalacofilho** [comment]
@jcward, couldn't verify it on my Ubuntu box.

You can try checking the syscalls with "strace" (`strace neko main.n`)?

**Comment by jcward** [comment]
Fascinating, so it worked for you... result of `strace neko test.n`:

```
write(1, "Main.hx:29: Move to /tmp folder,"..., 41Main.hx:29: Move to /tmp folder, fails...) = 41
write(1, "\n", 1
)                       = 1
rename("/home/jward/test.bin.2", "/tmp/move.bin") = -1 EXDEV (Invalid cross-device link)
stat("/tmp/move.bin", 0x7ffffc1d1090)   = -1 ENOENT (No such file or directory)
write(1, "Main.hx:31: FAIL!", 17Main.hx:31: FAIL!)       = 17
```

**Comment by jonasmalacofilho** [comment]
Ah!! Got it (from `man 2 rename`):

>       EXDEV  oldpath and newpath are not on the same mounted filesystem.  (Linux permits a filesystem to be mounted at multiple
              points, but rename() does not work across different mount points, even if the same filesystem is mounted on both.)

**Comment by jcward** [comment]
Ah ha, but as an app developer, I don't care how the filesystems are mounted, so how do we make it work? Are there alternatives to `rename`?

**Comment by jcward** [comment]
Huh, apparently this is a common issue -- a move utility method is often created that copies, verifies, and removes the original: http://stackoverflow.com/questions/7087743/how-to-rename-a-file-to-another-file-system

I suppose I'm proposing a `FileSystem.moveFile` method, or minimally documenting this shortcoming (not that I would have read that in advance!)

**Comment by jonasmalacofilho** [comment]
I'm actually looking into this now (never knew about `EXDEV` before), and it seems to be a huge mess.

The basic idea is indeed to copy the data _and metadata_ yourself and then ~~delete~~ `unlink` the old path.

However, the code for GNU's `mv` is much complex than that...  Check out the implementation of the [`copy` function](http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/copy.c#n2892) used by our beloved `mv`.

**Comment by jcward** [comment]
Ah metadata. Everything's more complex than one first imagines. :)

**Comment by back2dos** [comment]
I'm not really sure putting this into the standard library is the best idea. People should be aware of what's happening. A file moving operation such as the one you propose is a complex and delicate thing, where each of the many steps can go wrong, while a rename is pretty much just one atomic call.

IIRC there are a few file system libraries, that deal with these higher level operations. Not necessarily moving across file system boundaries, but recursive copying and deleting and searching and what not. That would be a better place for such functionality.

**Comment by jcward** [comment]
FYI, Apple man pages also list this restriction.

**Comment by jcward** [comment]
It sounds like a big complex operation... but apps cannot know what filesystem's are mounted where (you can mount anything/anywhere on both Linux and Mac, and perhaps even iOS and Android) which basically renders rename unreliable for any purpose other than renaming the file in the same folder. I suppose that's fine, but it's a fairly irritating gotcha that could at least be documented.

**Comment by jonasmalacofilho** [comment]
One thing that is particularly hard to deal with (in production) is that the exception thrown retains no information of the cause for the failure.

Maybe we could benefit from more detailed errors in the entire `sys.FileSystem` API...  They could at least include the corresponding syscall error code.

**Comment by ncannasse** [comment]
I agree we should not try to outsmart rename and keep it with the current limitation.
Could you submit a patch for neko/cpp that does handle such error (EXDEV) and throw an appropriate error?
